File: scripts/generate_translations.py
"""
BhuDhan Translation Generator

This script generates translation files for the BhuDhan application by translating
English strings to Hindi, Punjabi, Tamil, and Telugu using the translator module.

Usage:
    python generate_translations.py --input translations/en.json --output translations/

Requirements:
    pip install torch transformers sentencepiece sacremoses
"""

import argparse
import os
import json
import sys
from typing import Dict, List
from translator import translate_json, LANGUAGE_CODES

def ensure_dir(directory):
    """Ensure that a directory exists."""
    if not os.path.exists(directory):
        os.makedirs(directory)

def load_english_strings(input_file: str) -> Dict:
    """Load English strings from a JSON file."""
    try:
        with open(input_file, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as e:
        print(f"Error loading English strings: {e}")
        sys.exit(1)

def generate_translations(english_strings: Dict, output_dir: str, languages: List[str]):
    """Generate translation files for each language."""
    ensure_dir(output_dir)
    
    # Save the original English strings
    with open(os.path.join(output_dir, 'en.json'), 'w', encoding='utf-8') as f:
        json.dump(english_strings, f, ensure_ascii=False, indent=2)
    
    # Translate to each language
    for lang_code in languages:
        if lang_code == 'en':
            continue
            
        print(f"Translating to {LANGUAGE_CODES[lang_code]}...")
        translated_strings = translate_json(english_strings, 'en', lang_code)
        
        # Save translated strings
        output_file = os.path.join(output_dir, f"{lang_code}.json")
        with open(output_file, 'w', encoding='utf-8') as f:
            json.dump(translated_strings, f, ensure_ascii=False, indent=2)
        
        print(f"Saved {LANGUAGE_CODES[lang_code]} translations to {output_file}")

def main():
    parser = argparse.ArgumentParser(description="Generate translation files for BhuDhan application")
    
    # Define command-line arguments
    parser.add_argument("--input", type=str, required=True, 
                        help="Input JSON file with English strings")
    parser.add_argument("--output", type=str, default="translations", 
                        help="Output directory for translation files")
    parser.add_argument("--languages", type=str, default="hi,pa,ta,te", 
                        help="Comma-separated list of language codes to generate")
    
    args = parser.parse_args()
    
    # Parse languages
    languages = ['en'] + [lang.strip() for lang in args.languages.split(',')]
    
    # Validate languages
    for lang in languages:
        if lang not in LANGUAGE_CODES:
            print(f"Error: Unsupported language code '{lang}'")
            print(f"Supported languages: {', '.join(LANGUAGE_CODES.keys())}")
            sys.exit(1)
    
    # Load English strings
    english_strings = load_english_strings(args.input)
    
    # Generate translations
    generate_translations(english_strings, args.output, languages)
    
    print("Translation generation complete!")

if __name__ == "__main__":
    main()

File: scripts/generate-translations.js
#!/usr/bin/env node

/**
 * BhuDhan Translation Generator CLI
 * 
 * This script provides a command-line interface for generating translations
 * for the BhuDhan application.
 * 
 * Usage:
 *   node scripts/generate-translations.js
 */

const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

// Check if Python is installed
function checkPythonInstalled() {
  return new Promise((resolve, reject) => {
    const pythonProcess = spawn('python', ['--version']);
    
    pythonProcess.on('error', (error) => {
      reject(new Error('Python is not installed or not in PATH'));
    });
    
    pythonProcess.on('close', (code) => {
      if (code === 0) {
        resolve(true);
      } else {
        reject(new Error('Python is not installed or not in PATH'));
      }
    });
  });
}

// Check if required Python packages are installed
function checkRequiredPackages() {
  return new Promise((resolve, reject) => {
    const packages = ['torch', 'transformers', 'sentencepiece', 'sacremoses'];
    const pythonProcess = spawn('python', ['-c', `import sys; packages = ${JSON.stringify(packages)}; missing = [p for p in packages if p not in sys.modules and __import__(p, fromlist=['']) is None]; print(','.join(missing) if missing else '')`]);
    
    let output = '';
    
    pythonProcess.stdout.on('data', (data) => {
      output += data.toString();
    });
    
    pythonProcess.on('error', (error) => {
      reject(new Error('Failed to check required packages'));
    });
    
    pythonProcess.on('close', (code) => {
      if (code === 0) {
        const missingPackages = output.trim().split(',').filter(Boolean);
        if (missingPackages.length > 0) {
          reject(new Error(`Missing required Python packages: ${missingPackages.join(', ')}`));
        } else {
          resolve(true);
        }
      } else {
        reject(new Error('Failed to check required packages'));
      }
    });
  });
}

// Install required Python packages
function installRequiredPackages() {
  return new Promise((resolve, reject) => {
    console.log('Installing required Python packages...');
    
    const packages = ['torch', 'transformers', 'sentencepiece', 'sacremoses'];
    const pythonProcess = spawn('pip', ['install', ...packages]);
    
    pythonProcess.stdout.on('data', (data) => {
      console.log(data.toString());
    });
    
    pythonProcess.stderr.on('data', (data) => {
      console.error(data.toString());
    });
    
    pythonProcess.on('error', (error) => {
      reject(new Error(`Failed to install required packages: ${error.message}`));
    });
    
    pythonProcess.on('close', (code) => {
      if (code === 0) {
        console.log('Required packages installed successfully');
        resolve(true);
      } else {
        reject(new Error('Failed to install required packages'));
      }
    });
  });
}

// Generate translations
function generateTranslations() {
  return new Promise((resolve, reject) => {
    console.log('Generating translations...');
    
    // Path to the Python script
    const scriptPath = path.join(__dirname, 'generate_translations.py');
    
    // Input and output paths
    const inputPath = path.join(__dirname, '..', 'translations', 'en.json');
    const outputPath = path.join(__dirname, '..', 'translations');
    
    // Ensure the translations directory exists
    if (!fs.existsSync(outputPath)) {
      fs.mkdirSync(outputPath, { recursive: true });
    }
    
    // Spawn Python process
    const pythonProcess = spawn('python', [
      scriptPath,
      '--input', inputPath,
      '--output', outputPath
    ]);
    
    pythonProcess.stdout.on('data', (data) => {
      console.log(data.toString());
    });
    
    pythonProcess.stderr.on('data', (data) => {
      console.error(data.toString());
    });
    
    pythonProcess.on('error', (error) => {
      reject(new Error(`Failed to generate translations: ${error.message}`));
    });
    
    pythonProcess.on('close', (code) => {
      if (code === 0) {
        console.log('Translations generated successfully');
        resolve(true);
      } else {
        reject(new Error('Failed to generate translations'));
      }
    });
  });
}

// Main function
async function main() {
  try {
    // Check if Python is installed
    await checkPythonInstalled();
    console.log('Python is installed');
    
    try {
      // Check if required packages are installed
      await checkRequiredPackages();
      console.log('All required Python packages are installed');
    } catch (error) {
      console.error(error.message);
      
      // Install required packages
      await installRequiredPackages();
    }
    
    // Generate translations
    await generateTranslations();
    
    console.log('Translation generation completed successfully');
  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

// Run the main function
main();
